File: python/tests/exercises_test_.py
from dataclasses import FrozenInstanceError
import re
import math
import pytest
from exercises import (
    change, stretched, powers, say, find_first_then_lower,
    top_ten_scorers, crypto_functions, Quaternion)


def test_change():
    assert change(0) == (0, 0, 0, 0)
    assert change(97) == (3, 2, 0, 2)
    assert change(8) == (0, 0, 1, 3)
    assert change(250) == (10, 0, 0, 0)
    assert change(144) == (5, 1, 1, 4)
    assert change(97) == (3, 2, 0, 2)
    assert change(100000000000) == (4000000000, 0, 0, 0)
    with pytest.raises(TypeError) as excinfo:
        # No fractional amounts allowed
        change(988.25)
    with pytest.raises(TypeError) as excinfo:
        # Do type check before negative check
        change(-988.25)
    with pytest.raises(ValueError) as excinfo:
        change(-50)
    assert str(excinfo.value) == 'amount cannot be negative'


def test_stretched():
    assert stretched('') == ''
    assert stretched('H e   l\t\tlo') == 'Heelllllllooooo'
    assert stretched('$#') == '$##'
    assert stretched('       ') == ''
    assert stretched('A = πr²') == 'A==πππrrrr²²²²²'
    assert stretched("  ") == ""
    assert stretched("  \t\n  \t") == ""
    assert stretched("  Hi  hi  ") == "Hiihhhiiii"
    assert stretched("😁😂😱") == "😁😂😂😱😱😱"
    assert stretched("hello world") == "heelllllllooooowwwwwwooooooorrrrrrrrllllllllldddddddddd"
    with pytest.raises(TypeError):
        # Test that the argument MUST be a positional argument
        stretched(s = '')


def test_powers():
    p = powers(base=2, limit=10)
    assert next(p) == 1
    assert next(p) == 2
    assert next(p) == 4
    assert next(p) == 8
    with pytest.raises(StopIteration):
        next(p)
    assert list(powers(base=2, limit=-5)) == []
    assert list(powers(base=7, limit=0)) == []
    assert list(powers(limit=1, base=3)) == [1]
    assert list(powers(base=2, limit=63)) == [1, 2, 4, 8, 16, 32]
    assert list(powers(base=2, limit=64)) == [1, 2, 4, 8, 16, 32, 64]
    with pytest.raises(TypeError):
        # Test that the arguments MUST be keyword arguments
        powers(2, 100)


def test_say():
    assert say() == ''
    assert say('hi')() == 'hi'
    assert say('hi')('there')() == 'hi there'
    assert say('hello')('my')('name')('is')(
        'Colette')() == 'hello my name is Colette'
    with pytest.raises(TypeError):
        # Test that the argument MUST be a positional argument
        say(s = '')


def test_find_first_then_lower_case():
    with pytest.raises(ValueError):
        find_first_then_lower(lambda s: len(s) > 10, [])
    with pytest.raises(ValueError):
        find_first_then_lower(lambda s: len(s) > 5, ["hello", "world"])
    assert find_first_then_lower(lambda s: s.startswith("HELL"), ["HELLO", "WORLD"]) == "hello"
    assert find_first_then_lower(lambda s: "d!" in s, ["Hello", "World!!"]) == "world!!"


def test_top_ten_scorers():
    assert top_ten_scorers({}) == []
    assert top_ten_scorers({'T1': [['A', 3, 300]]}) == []
    assert top_ten_scorers({'T1': [['A', 30, 300]]}) == ['A|10.00|T1']
    input = {
        'ATL': [
            ['Betnijah Laney', 16, 263],
            ['Courtney Williams', 14, 193]],
        'CHI': [
            ['Kahleah Copper', 17, 267],
            ['Allie Quigley', 17, 260],
            ['Courtney Vandersloot', 17, 225]],
        'CONN': [
            ['DeWanna Bonner', 16, 285],
            ['Alyssa Thomas', 16, 241]],
        'DAL': [
            ['Arike Ogunbowale', 16, 352],
            ['Satou Sabally', 12, 153]],
        'IND': [
            ['Kelsey Mitchell', 16, 280],
            ['Tiffany Mitchell', 13, 172],
            ['Candice Dupree', 16, 202]],
        'LA': [
            ['Nneka Ogwumike', 14, 172],
            ['Chelsea Gray', 16, 224],
            ['Candace Parker', 16, 211]],
        'LV': [
            ['A’ja Wilson', 15, 304],
            ['Dearica Hamby', 15, 188],
            ['Angel McCoughtry', 15, 220]],
        'MIN': [
            ['Napheesa Collier', 16, 262],
            ['Crystal Dangerfield', 16, 254]],
        'NY': [
            ['Layshia Clarendon', 15, 188]],
        'PHX': [
            ['Diana Taurasi', 13, 236],
            ['Brittney Griner', 12, 212],
            ['Skylar Diggins-Smith', 16, 261],
            ['Bria Hartley', 13, 190]],
        'SEA': [
            ['Breanna Stewart', 16, 317],
            ['Jewell Loyd', 16, 223]],
        'WSH': [
            ['Emma Meesseman', 13, 158],
            ['Ariel Atkins', 15, 212],
            ['Myisha Hines-Allen', 15, 236]]}
    expected = [
        "Arike Ogunbowale|22.00|DAL",
        "A’ja Wilson|20.27|LV",
        "Breanna Stewart|19.81|SEA",
        "DeWanna Bonner|17.81|CONN",
        "Kelsey Mitchell|17.50|IND",
        "Betnijah Laney|16.44|ATL",
        "Napheesa Collier|16.38|MIN",
        "Skylar Diggins-Smith|16.31|PHX",
        "Crystal Dangerfield|15.88|MIN",
        "Myisha Hines-Allen|15.73|WSH"
    ]
    assert top_ten_scorers(input) == expected


def test_crypto_functions():
    encode, decode = crypto_functions()
    for message in (b'', b'abc', bytes('😜😍😞🦾👀', 'utf-8')):
        assert(decode(encode(message))) == message


def test_quaternions():
    q1 = Quaternion(1, -2, 5, 0)
    q2 = Quaternion(3, 0, 5, -2)
    assert (q1 + q2).coefficients == (4, -2, 10, -2)
    assert (q1 * q2).coefficients == (-22, -16, 16, -12)
    with pytest.raises(FrozenInstanceError):
        # Test immutability
        q2.a = 100

File: java/ExercisesTest.java
import org.junit.jupiter.api.Test;

import static org.junit.jupiter.api.Assertions.assertArrayEquals;
import static org.junit.jupiter.api.Assertions.assertEquals;
import static org.junit.jupiter.api.Assertions.assertThrows;

import java.util.List;
import java.util.Map;
import java.util.HashMap;
import java.util.Optional;
import java.util.function.IntConsumer;
import java.util.*;

public class ExercisesTest {

    @Test
    public void testChange() {
        assertThrows(IllegalArgumentException.class, () -> Exercises.change(-20));
        assertEquals(List.of(0, 0, 0, 0), Exercises.change(0));
        assertEquals(List.of(0, 0, 0, 1), Exercises.change(1));
        assertEquals(List.of(0, 1, 0, 3), Exercises.change(13));
        assertEquals(List.of(2, 0, 0, 0), Exercises.change(50));
        assertEquals(List.of(0, 0, 1, 0), Exercises.change(5));
        assertEquals(List.of(15, 2, 0, 2), Exercises.change(397));
        assertEquals(List.of(1, 1, 1, 2), Exercises.change(42));
        assertEquals(List.of(40000000, 1, 1, 2), Exercises.change(1000000017));
    }

    @Test
    public void testStretched() {
        assertEquals("", Exercises.stretched(""));
        assertEquals("", Exercises.stretched("  "));
        assertEquals("", Exercises.stretched("  \t\n  \t"));
        assertEquals("Hiihhhiiii", Exercises.stretched("  Hi  hi  "));
        assertEquals("😁😂😂😱😱😱", Exercises.stretched("😁😂😱"));
        assertEquals(
                "heelllllllooooowwwwwwooooooorrrrrrrrllllllllldddddddddd",
                Exercises.stretched("hello world"));
    }

    @Test
    public void testPowersWithConsumer() {
        var a = new ArrayList<Integer>();
        IntConsumer addToList = i -> a.add(i);
        Exercises.powers(2, 1, addToList);
        assertEquals(List.of(1), a);
        Exercises.powers(-3, 81, addToList);
        assertEquals(List.of(1, 1, -3, 9, -27, 81, -243), a);
    }

    @Test
    public void testPowersStream() {
        assertArrayEquals(
                new int[] { 1, 1, 1, 1, 1 },
                Exercises.powersStream(1).limit(5).toArray());
        assertArrayEquals(Exercises.powersStream(7).limit(10).toArray(),
                new int[] { 1, 7, 49, 343, 2401, 16807, 117649, 823543,
                        5764801,
                        40353607 });
        assertArrayEquals(Exercises.powersStream(-3).limit(5).toArray(),
                new int[] { 1, -3, 9, -27, 81 });
        assertArrayEquals(Exercises.powersStream(10).limit(4).toArray(),
                new int[] { 1, 10, 100, 1000 });
    }

    @Test
    public void testSay() {
        assertEquals("A", Exercises.say("A").ok());
        assertEquals("", Exercises.say());
        assertEquals("A B", Exercises.say("A").and("B").ok());
        assertEquals("🐤🦇 $🦊👏🏽 !",
                Exercises.say("🐤🦇").and("$🦊👏🏽").and("!").ok());
    }

    @Test
    public void testFindFirstAndLowerCase() {
        assertEquals(Optional.empty(),
                Exercises.findFirstThenLower(s -> s.length() > 10,
                        List.of()));
        assertEquals(Optional.empty(),
                Exercises.findFirstThenLower(s -> s.length() > 5,
                        List.of("hello", "world")));
        assertEquals(Optional.of("hello"),
                Exercises.findFirstThenLower(s -> s.startsWith("HELL"),
                        List.of("HELLO", "WORLD")));
        assertEquals(Optional.of("world!!"),
                Exercises.findFirstThenLower(s -> s.contains("d!"),
                        List.of("Hello", "World!!")));
    }

    @Test
    public void testTopTenScorers() {
        assertEquals(List.of(), Exercises.topTenScorers(Map.of()));

        assertEquals(List.of("A|10.00|T1"),
                Exercises.topTenScorers(
                        Map.of("T1", List.of("A,30,300"))));

        var stats = new HashMap<String, List<String>>();
        stats.put("ATL", List.of(
                "Betnijah Laney,16,263", "Courtney Williams,14,193"));
        stats.put("CHI", List.of(
                "Kahleah Copper,17,267", "Allie Quigley,17,260",
                "Courtney Vandersloot,17,225"));
        stats.put("CONN", List.of(
                "DeWanna Bonner,16,285", "Alyssa Thomas,16,241"));
        stats.put("DAL", List.of(
                "Arike Ogunbowale,16,352", "Satou Sabally,12,153"));
        stats.put("IND", List.of(
                "Kelsey Mitchell,16,280", "Tiffany Mitchell,13,172",
                "Candice Dupree,16,202"));
        stats.put("LA", List.of(
                "Nneka Ogwumike,14,172", "Chelsea Gray,16,224",
                "Candace Parker,16,211"));
        stats.put("LV", List.of(
                "A’ja Wilson,15,304", "Dearica Hamby,15,188",
                "Angel McCoughtry,15,220"));
        stats.put("MIN", List.of(
                "Napheesa Collier,16,262", "Crystal Dangerfield,16,254"));
        stats.put("NY", List.of(
                "Layshia Clarendon,15,18"));
        stats.put("PHX", List.of(
                "Diana Taurasi,13,236", "Brittney Griner,12,212",
                "Skylar Diggins-Smith,16,261",
                "Bria Hartley,13,190"));
        stats.put("SEA", List.of(
                "Breanna Stewart,16,317", "Jewell Loyd,16,223"));
        stats.put("WSH", List.of(
                "Emma Meesseman,13,158", "Ariel Atkins,15,212",
                "Myisha Hines-Allen,15,236"));

        assertEquals(
                List.of(
                        "Arike Ogunbowale|22.00|DAL",
                        "A’ja Wilson|20.27|LV",
                        "Breanna Stewart|19.81|SEA",
                        "DeWanna Bonner|17.81|CONN",
                        "Kelsey Mitchell|17.50|IND",
                        "Betnijah Laney|16.44|ATL",
                        "Napheesa Collier|16.38|MIN",
                        "Skylar Diggins-Smith|16.31|PHX",
                        "Crystal Dangerfield|15.88|MIN",
                        "Myisha Hines-Allen|15.73|WSH"),
                Exercises.topTenScorers(stats));
    }

    @Test
    public void testQuaternionConstructorErrors() {
        // That that NaNs are detected in each argument
        assertThrows(IllegalArgumentException.class,
                () -> new Quaternion(Double.NaN, 0, 0, 0));
        assertThrows(IllegalArgumentException.class,
                () -> new Quaternion(0, Double.NaN, 0, 0));
        assertThrows(IllegalArgumentException.class,
                () -> new Quaternion(0, 0, Double.NaN, 0));
        assertThrows(IllegalArgumentException.class,
                () -> new Quaternion(0, 0, 0, Double.NaN));
    }

    @Test
    public void testQuaternionAccessors() {
        // Java records give us these for free
        var q = new Quaternion(3.5, 2.25, -100, -1.25);
        assertEquals(3.5, q.a());
        assertEquals(2.25, q.b());
        assertEquals(-100.0, q.c());
        assertEquals(-1.25, q.d());
    }

    @Test
    public void testQuaternionArithmetic() {
        var q1 = new Quaternion(1, 3, 5, 2);
        var q2 = new Quaternion(-2, 2, 8, -1);
        var q3 = new Quaternion(-1, 5, 13, 1);
        var q4 = new Quaternion(-46, -25, 5, 9);
        assertEquals(q3, q1.plus(q2));
        assertEquals(q1, q3.minus(q2));
        assertEquals(q4, q1.times(q2));
        assertEquals(Quaternion.K, Quaternion.I.times(Quaternion.J));
    }

    @Test
    public void testQuaternionCoefficients() {
        assertEquals(List.of(0.0, 0.0, 0.0, 0.0),
                new Quaternion(0, 0, 0, 0).coefficients());
        assertEquals(List.of(2.0, 1.5, 10.0, -8.0),
                new Quaternion(2, 1.5, 10, -8).coefficients());
    }

    @Test
    public void testQuaternionToString() {
        assertEquals("Quaternion[a=0.0, b=0.0, c=0.0, d=0.0]",
                new Quaternion(0, 0, 0, 0).toString());
        assertEquals("Quaternion[a=0.0, b=-1.0, c=0.0, d=2.25]",
                new Quaternion(0, -1, 0, 2.25).toString());
        assertEquals("Quaternion[a=0.0, b=0.0, c=0.0, d=-1.0]",
                Quaternion.ZERO.minus(Quaternion.K).toString());
    }
}